# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmInstance.h
# Commit: 1a73cfde72e8
# Full Hash: 1a73cfde72e8f4d19bdcfdbffa9fbc78bba01e24
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2021-09-28 09:54:33
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Introduce indirect stubs to optimize call_indirect. r=lth
#   
#   This patch introduces indirect stubs.
#   An indirect stub is a stub that takes care of any switching activities needed for call_indirect.
#   Before this patch, we have to conservatively assume that any call_indirect's target can be from a foreign instance,
# ==============================================================================

diff -r da3c5d8492db -r 1a73cfde72e8 js/src/wasm/WasmInstance.h
--- a/js/src/wasm/WasmInstance.h	Tue Sep 28 05:16:23 2021 +0000
+++ b/js/src/wasm/WasmInstance.h	Tue Sep 28 05:24:15 2021 +0000
@@ -66,6 +66,9 @@
   const void** addressOfTypeId(const TypeIdDesc& typeId) const;
   FuncImportTls& funcImportTls(const FuncImport& fi);
   TableTls& tableTls(const TableDesc& td) const;
+  void* checkedCallEntry(const uint32_t functionIndex, const Tier tier) const;
+  void* checkedCallEntryForWasmImportedFunction(JSFunction* fun,
+                                                const Tier tier) const;
 
   // Only WasmInstanceObject can call the private trace function.
   friend class js::WasmInstanceObject;
@@ -98,6 +101,8 @@
                        uint8_t* nextPC,
                        uintptr_t highestByteVisitedInPrevFrame);
 
+  Instance* getOriginalInstanceAndFunction(uint32_t funcIdx, JSFunction** fun);
+
   JS::Realm* realm() const { return realm_; }
   const Code& code() const { return *code_; }
   const CodeTier& code(Tier t) const { return code_->codeTier(t); }
@@ -173,6 +178,17 @@
                                uint32_t dstOffset, uint32_t srcOffset,
                                uint32_t len);
 
+  void* getIndirectStub(uint32_t funcIndex, TlsData* targetTlsData,
+                        const Tier tier) const;
+  void* createIndirectStub(uint32_t funcIndex);
+  bool createManyIndirectStubs(const VectorOfIndirectStubTarget& targets,
+                               const Tier tier);
+  bool ensureIndirectStubs(const Uint32Vector& elemFuncIndices,
+                           uint32_t srcOffset, uint32_t len, const Tier tier,
+                           const bool tableIsImportedOrExported);
+  bool ensureIndirectStub(FuncRef* ref, const Tier tier,
+                          const bool tableIsImportedOrExported);
+
   // Debugger support:
 
   JSString* createDisplayURL(JSContext* cx);